### PR TITLE
EAHW1249 (annual target hours)

### PR DESCRIPTION
**Pull request by wweatherill-homeoffice**

### CODE REVIEW

**Review by dawatkin3 — DISMISSED** [review]

**Review by dawatkin3 — APPROVED** [review]

File: docs/containers/rest-api/index.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+None
+
+## Container contract
+The container exposes its functionality through a set of RESTful endpoints that are documented in the [rest-endpoints.md](./rest-endpoints.md) page
+
+## Contents
+
+- [container-model.c4](./container-model.c4) - a C4 model of the system, used to generate images that visualise the API of the container and components within it
+- [rest-endpoints.md](./rest-endpoints.md) - generated from [openapi.yaml](../../../src/main/resources/openapi.yaml) it describes the REST endpoints
+- [rest-operations.md](./rest-operations.md) - some of the resources exposed by the REST API are calculated. This page list those resources and also describes how they are calculated
```
Comment (laurieboyes) [inline docs/containers/rest-api/index.md:21]: dead link

File: docs/containers/balance-calculator/orchestration/accrual-balance-calculation.md [hunk 2]
```diff
@@ -0,0 +1,565 @@
...
+      - [Scenario 3: Move an existing `TimeEntry` spanning a single day to a different time on the same day](#scenario-3--move-an-existing--timeentry--spanning-a-single-day-to-a-different-time-on-the-same-day)
+      - [Scenario 4: Create a `TimeEntry` spanning two contiguous days](#scenario-4--create-a--timeentry--spanning-two-contiguous-days)
+      - [Scenario 5: Move an existing `TimeEntry` spanning a single day to two contiguous days](#scenario-5--move-an-existing--timeentry--spanning-a-single-day-to-two-contiguous-days)
+      - [Scenario 6: Move an existing `TimeEntry` spanning two contiguous days a single day](#scenario-6--move-an-existing--timeentry--spanning-two-contiguous-days-a-single-day)
+      - [Scenario 7: Delete an existing `TimeEntry` spanning two contiguous days](#scenario-7--delete-an-existing--timeentry--spanning-two-contiguous-days)
+      - [Scenario 8: Move an existing `TimeEntry` spanning a single day to a different single day](#scenario-8--move-an-existing--timeentry--spanning-a-single-day-to-a-different-single-day)
```
Comment (laurieboyes) [inline docs/containers/balance-calculator/orchestration/accrual-balance-calculation.md:21]: I think these links have broken

File: docs/containers/balance-calculator/orchestration/accrual-balance-calculation.md [hunk 3]
```diff
@@ -0,0 +1,565 @@
...
+## How to calculate `contribution.value`
+When a new `TimeEntry` is received then Accruals must recalculate effected balances. Depending on how many contiguous days the new `TimeEntry` covers and whether or not the `TimeEntry` represents an update (typically to move time around) or a deletion then the action required is different.
+
+### Single day
+If the date portions of the `TimeEntry.actualStartTime` and `TimeEntry.actualEndTime` are for the same date then this `TimeEntry` covers a single day.
+
+`value` = `TimeEntry.actualEndTime` - `TimeEntry.actualStartTime`
+
+### Overlapping day
+If the date portions of the `TimeEntry.actualStartTime` and `TimeEntry.actualEndTime` are for the different dates then this `TimeEntry` covers multiple days. The assumption is that the days are contiguous and that typically no more than two days are spanned.
```
Comment (laurieboyes) [inline docs/containers/balance-calculator/orchestration/accrual-balance-calculation.md:44]: It’s definitely implied by the new GMT to BST section (_sick_ diagrams btw), but is it worth having a super clear statement here saying words to the effect of ‘when apportioning accruals to days for overlapping shifts we use UK time (inclusive of daylight savings)’

Apologies if this already exists somewhere and I’ve missed it

File: docs/containers/balance-calculator/orchestration/index.md [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+Accrual instances are found via a RESTful call to the `callisto-accrual-restapi` component (`GET` [/resources/accruals](./../../rest-api/rest-endpoints.md#findaccruals))
+
+#### Balance calculator
+Having found the Accrual instances which are to be updated this component is responsible for calculating new balances and updating the owning Accrual instance. More information on how to use TimeCard event data to calculate a balance can be found in [accrual-balance-calculation.md](./accrual-balance-calculation.md).
+
+### Atomic transactions
+All Accruals resources that are passed to the `accuals-rest-api`'s batch update endpoint should be committed in a single transaction.
+
+In the design presented here all Accrual resources are sent in a single call to the batch update endpoint.
```
Comment (laurieboyes) [inline docs/containers/balance-calculator/orchestration/index.md:89]: Is there any reason for the design to be agnostic regarding whether the accruals updates for all types are all committed in parallel or not? Could the design make it clear that only accruals of the same type need to be committed in a transaction?

If I’m reading right, it seems like the design leaves this up to engineers, but as it reaches between containers, to me it feels like something on which it’s worth giving architectural steer

File: docs/containers/balance-calculator/container-model.c4 [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+            timecardEventsProducer -> accrualsBalanceCalculator "consumes TimeCard event from"
+            accrualsBalanceCalculator -> accrualsBalanceCalculator "identifies Accrual instannces and calculates balances"
+            accrualsBalanceCalculator -> accrualsRestApi "stores updated Accrual instance using"
+        }
+        
+        dynamic accrualsBalanceCalculator "calculateAndUpdateAccrualBalancesDetail" "Consumption of TimeCard event updates one or more Accrual instances" {        
+            timecardEventsProducer -> timecardEventsConsumer "consumes TimeCard event from"
+            timecardEventsConsumer -> orchestrator "passes TimeCard event to"
+            orchestrator -> accrualTypeIdentifier "determines if TimeCard event is relevant to Accrual type using"
+            accrualTypeIdentifier -> accrualsRestApi "Finds Accrual types effected by TimeCard event using"
```
Comment (laurieboyes) [inline docs/containers/balance-calculator/container-model.c4:50]:
```suggestion
            accrualTypeIdentifier -> accrualsRestApi "Finds Accrual types affected by TimeCard event using"
```

File: docs/containers/rest-api/storage.md [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+
+A contribution towards the balance of an Accrual
+
+### Properties
+
+|Name|Type|Required|Restrictions|Description|
+|---|---|---|---|---|
+|id|PK|true|none|the identifier for this Contribution|
+|time_entry_id|FK ([time_entry](#tocS_TimeEntry))|true|none|The TimeEntry records an amount of time that backs the contributedHours property|
+|accrual_id|FK ([accrual](#tocS_Accrual))|true|none|The balance of the `Accrual` that this `Contribution` effects|
```
Comment (laurieboyes) [inline docs/containers/rest-api/storage.md:44]:
```suggestion
|accrual_id|FK ([accrual](#tocS_Accrual))|true|none|The balance of the `Accrual` that this `Contribution` affects|
```

File: docs/containers/balance-calculator/orchestration/index.md [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+
+Since the process outlined above has a dependency on the `callisto-accrual-restapi` it is possible that there could be a persistent failure to complete the process.
+
+The implementation should guard against this by using a combination of an [exponential back off retry for transitive errors and a circuit breaker for handling persistent failure](https://dzone.com/articles/understanding-retry-pattern-with-exponential-back)
+
+The algorithm for determining which Accrual type a TimeCard event should effect varies by Accrual type. On this basis it is envisaged that there will be a series of concrete implementations of an Accrual Type identifier interface and the Orchestrator simply cycles through each asking if the TimeCard event ties to the given Accrual type that the implementation knows about.
```
Comment (laurieboyes) [inline docs/containers/balance-calculator/orchestration/index.md:74]:
```suggestion
The algorithm for determining which Accrual type a TimeCard event should affect varies by Accrual type. On this basis it is envisaged that there will be a series of concrete implementations of an Accrual Type identifier interface and the Orchestrator simply cycles through each asking if the TimeCard event ties to the given Accrual type that the implementation knows about.
```

File: docs/containers/balance-calculator/orchestration/index.md [hunk 8]
```diff
@@ -0,0 +1,100 @@
+# Accrual Balance maintenance
+
+End users do not directly make changes to Accruals data. Instead the balance-calculator component responds to events produced by the TimeCard container.
+
+There are two main activities initiated in TimeCard that Accruals is interested in
+
+### 1. User records their time
+
+When a user records time in their TimeCard then the Accruals container needs to respond by updating the balances with those Accrual modules that are effected by the newly recorded time.
```
Comment (laurieboyes) [inline docs/containers/balance-calculator/orchestration/index.md:9]:
```suggestion
When a user records time in their TimeCard then the Accruals container needs to respond by updating the balances with those Accrual modules that are affected by the newly recorded time.
```

File: docs/containers/balance-calculator/orchestration/accrual-balance-calculation.md (hunk 2 above)
Comment (wweatherill-homeoffice) [reply docs/containers/balance-calculator/orchestration/accrual-balance-calculation.md]: Links now changed and tested

File: docs/containers/balance-calculator/orchestration/accrual-balance-calculation.md (hunk 3 above)
Comment (wweatherill-homeoffice) [reply docs/containers/balance-calculator/orchestration/accrual-balance-calculation.md:44]: Addressed under the GMT to BST section - now linking out to a Confluence page written by @stephenkaye-ho around the use of ISO 8061

File: docs/containers/balance-calculator/orchestration/index.md (hunk 4 above)
Comment (wweatherill-homeoffice) [reply docs/containers/balance-calculator/orchestration/index.md:89]: We didn't want to optimise prematurely. At the moment I don't think we really have a good feel for whether or not one large update is "better" than several smaller updates. The simplest thing for now is to perform one large update hence that approach being the one followed by the design. 

I think we might need to revisit that decision later but the RESTful API should not need to change. Instead we'd just end up calling it more times as part of calculating the accrual balances for a given person. Granted the internals of the balance calculator would have to change to chunk the work up but as I say, at the moment this feels like a premature optimisation

**Review by taomoh — APPROVED** [review]

**Review by stephenkaye-ho — APPROVED** [review]

**Review by stephenkaye-ho — APPROVED** [review]

**Review by stephenkaye-ho — APPROVED** [review]

**Pull request merged**